pylint/checkers/variables.py

```diff
@@ -1781,20 +1781,18 @@ def _is_variable_violation(
             frame, nodes.FunctionDef
         ):
             # Special rule for function return annotations,
-            # which uses the same name as the class where
-            # the function lives.
+            # using a name defined earlier in the class containing the function.
             if node is frame.returns and defframe.parent_of(frame.returns):
-                maybe_before_assign = annotation_return = True
-
-            if (
-                maybe_before_assign
-                and defframe.name in defframe.locals
-                and defframe.locals[node.name][0].lineno < frame.lineno
-            ):
-                # Detect class assignments with the same
-                # name as the class. In this case, no warning
-                # should be raised.
-                maybe_before_assign = False
+                annotation_return = True
+                if (
+                    frame.returns.name in defframe.locals
```

Comment:
The only substantive change in the PR is to change `defframe.name` to `frame.returns.name`. The rest is just refactoring for readability and avoiding re-checking the same boolean `maybe_before_assign`.